base/abstractdict.jl

```diff
@@ -527,6 +527,71 @@ function hash(a::AbstractDict, h::UInt)
     hash(hv, h)
 end
 
+"""
+    modify!(f, d::AbstractDict{K, V}, key) -> (old => new)
+
+Lookup and then update, insert or delete in one go without re-computing the hash.
+Return a pair of `old::Union{Some,Nothing}` and `new::::Union{Some,Nothing}`.
+
+`f` is a callable object that must take a single `old::Union{Some, Nothing}` argument and
+return a `new::Union{T, Some{T}, Nothing}` value, where `T` is a type [`convert`](@ref)-able
+to the value type `V`.  The value `Some(d[key])` is passed to `f` if `haskey(d, key)`;
+otherwise `nothing` is passed.  If `f` returns `nothing`, corresponding entry in the
+dictionary `d`  is removed.  If `f` returns non-`nothing` value `x`, `key => something(x)`
+is inserted or updated in `d` (equivalent to `d[key] = something(x)` but more efficient).
+
+Whether `Some{V}(value)` or `Some{typeof(value)}(value)` is returned is an implementation
+defined behavior. The callback function `f` must use `old === nothing` or `old isa Some`
+instead of `old isa Some{valtype(d)}` unless the type of the dictionary `d` is known to
+define a certain behavior.
```

Comment:
It seems like a very rare event to know about a certain specific behavior of a dict type. I would just remove "unless ...", and change "must use" to   

> The callback function `f` should generally use `old === nothing` or `old isa Some` instead ...

(just a suggestion)

Reply:
I prefer requreing the dict to return `Some{V}`